py/dml/structure.py

```diff
@@ -608,22 +636,47 @@ def is_default(decl):
         else:
             raise EAMBINH(
                 p1.site, p2.site, name, rank1.desc, rank2.desc,
-                is_default((rank1, p1)))
+                decl_is_default((rank1, p1)))
+
+    decl_ranks = {rank for (rank, _) in decls}
+    decl_or_def_ranks = decl_ranks.union(rank for (rank, _) in defs)
+    for (rank, p) in defs:
+        if p.site.provisional_enabled(provisional.explicit_param_decls):
+            (_, type_info, is_default, _) = p.args
+            if type_info is None:
+                declared_as_override = True
+            else:
+                assert type_info.kind in {'walrus', 'paramtype'}
+                declared_as_override = False
+            is_override = (not decl_or_def_ranks.isdisjoint(rank.inferior)
+                           or rank in decl_ranks)
+            if not declared_as_override and is_override:
+                [parent, *_] = [
+                    other for (other_rank, other) in decls + defs
+                    if other_rank in rank.inferior] + [
+                    other for (other_rank, other) in decls
+                    if other_rank is rank]
+                report(EOVERRIDE(type_info.site, type_info.site,#parent.site,
+                               'default' if is_default else '='))
+            elif (not is_override and declared_as_override
+                  and not any(r is rank for (r, _) in decls)):
+                report(ENOVERRIDE(
+                    p.site, 'default' if is_default else '='))
```

Comment:
This entire loop strikes me as very messy. I think this could greatly benefit from the newly broken out `calc_minimal_ancestry`:
```
rank_to_decl_or_def = {r: spec for (r, spec) in itertools.chain(decls, defs) }
ancestry = calc_minimal_ancestry(Set(rank_to_decl_or_def))

for (rank, p) in defs:
    if p.site.provisional_enabled(provisional.explicit_param_decls):
        overrides = list(ancestry[rank])
        if not declared_as_override and overrides:
                report(EOVERRIDE(type_info.site, overrides[0], 'default' if is_default else '='))
        ....
```
... Actually, maybe the entirety of `merge_parameters` could be simplified by this.

Reply:
@mandolaerik I've seen your new code -- how come you are still clumsily determining a parent via a list comprehension? I don't get it. Why would you not use the same or a similar pattern as I exemplify in this comment?

Reply:
oh, maybe i'm confusing. In the above comment code: `s/overrides/overridden`. I meant in the sense "that which p overrides", but that is not a good choice of words

Reply:
The strongest reason is that `list(ancestry[rank])[0]` depends on set iteration order. Also, your suggestion reduces to something like:
```
parent = dict(itertools.chain(decls.items(), defs.items()))[next(iter(minimal_ancestry[rank]))]
```
which I find harder to read.

Reply:
With the code as-is you might as well not use `calc_minimal_ancestry` at all. I don't buy the "if don't like it cause if I minimize the code it becomes ugly" argument; then don't minimize the code. This does not affect generated C, so the undefined set iteration order argument is weak; but does carry the most weight.
What about reworking `Rank` to operate in terms of `Set`? Would the induced slowdown be too much? If so, if you insist on not using the values of the map `calc_minimal_ancestry` gives you, then don't use the function at all. I still feel the loop, both before and as-is (because as-is is just the exact same pattern & algorithm you employed before, just with `calc_minimal_ancestry` shoe-horned in) is _hideously_ ugly, and I would **really** want it improved, but if you can't find a way then so be it.

Reply:
Sorry, I don't see how your suggestion would improve the situation.
```
is_override = not all_ranks.isdisjoint(rank.inferior
```
vs
```
is_override = bool(calc_minimal_ancestry(all_ranks)[rank])
```
and
```
                minimal_ancestry = traits.calc_minimal_ancestry(all_ranks)
                [parent, *_] = (parent for (parent_rank, parent) in params
                                if parent_rank in minimal_ancestry[rank])
```
vs
```
                rank_to_decl_or_def = dict(itertools.chain(defs, decls))
                all_overrides = traits.calc_minimal_ancestry(all_ranks)
                [one_override, *] = all_overrides
                parent = rank_to_decl_or_def[one_override]
```
I don't understand why the latter would be better code.

Reply:
Because
```
rank_to_decl_or_def = dict(itertools.chain(defs, decls))
all_overrides = traits.calc_minimal_ancestry(all_ranks)
```
Doesn't have to be part of the loop? You don't have to, you know, _recalculate_ `calc_minimal_ancestry` for every param def/decl. And I find avoiding the need to clumsily spell out `(parent for (parent_rank, parent) in params if parent_rank in minimal_ancestry[rank])` a benefit.

Reply:
```
all_overrides = traits.calc_minimal_ancestry(all_ranks)
[one_override, *] = all_overrides
```
isn't even legal. Are you simply misunderstanding what `calc_minimal_ancestry` returns?

Reply:
In particular, your example of `is_override = bool(calc_minimal_ancestry(all_ranks)[rank])` is disingenuous because you can just have the result of `calc_minimal_ancestry(all_ranks)` be calculated before the loop and shared between each iteration of it, reducing the check to `is_override = bool(minimal_ancestry[rank])`.

Reply:
Things are somewhat complex, there are performance concerns. I solved it by memoizing calc_minimal_ancestry.